File: internal/commands/ping_test.go
package commands

import (
	"context"
	"errors"
	"strings"
	"testing"

	"github.com/gcerar/joplin-axi/internal/args"
	"github.com/gcerar/joplin-axi/internal/client/clienttest"
)

func TestPingCommand(t *testing.T) {
	ctx := context.Background()
	noArgs := args.ParsedArgs{}

	t.Run("reports unreachable and failed auth without attempting an authenticated call", func(t *testing.T) {
		stub := &clienttest.StubClient{
			PingFunc:          func() bool { return false },
			ListNotebooksFunc: func([]string) ([]map[string]any, error) { return nil, nil },
		}

		result, err := PingCommand.Run(ctx, noArgs, stub)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !strings.Contains(result.Output, "clipper: unreachable") {
			t.Errorf("output %q does not contain 'clipper: unreachable'", result.Output)
		}
		if !strings.Contains(result.Output, "auth: failed") {
			t.Errorf("output %q does not contain 'auth: failed'", result.Output)
		}
		if len(stub.ListNotebooksCalls) != 0 {
			t.Error("ListNotebooks was called despite the clipper being unreachable")
		}
	})

	t.Run("reports reachable and auth ok when an authenticated call succeeds", func(t *testing.T) {
		stub := &clienttest.StubClient{
			PingFunc:          func() bool { return true },
			ListNotebooksFunc: func([]string) ([]map[string]any, error) { return nil, nil },
		}

		result, err := PingCommand.Run(ctx, noArgs, stub)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !strings.Contains(result.Output, "clipper: reachable") {
			t.Errorf("output %q does not contain 'clipper: reachable'", result.Output)
		}
		if !strings.Contains(result.Output, "auth: ok") {
			t.Errorf("output %q does not contain 'auth: ok'", result.Output)
		}
	})

	t.Run("reports reachable but failed auth when the token is rejected", func(t *testing.T) {
		stub := &clienttest.StubClient{
			PingFunc:          func() bool { return true },
			ListNotebooksFunc: func([]string) ([]map[string]any, error) { return nil, errors.New("401") },
		}

		result, err := PingCommand.Run(ctx, noArgs, stub)
		if err != nil {
			t.Fatalf("unexpected error: %v", err)
		}
		if !strings.Contains(result.Output, "clipper: reachable") {
			t.Errorf("output %q does not contain 'clipper: reachable'", result.Output)
		}
		if !strings.Contains(result.Output, "auth: failed") {
			t.Errorf("output %q does not contain 'auth: failed'", result.Output)
		}
	})
}

File: test/ping.test.ts
import { describe, expect, it, vi } from 'vitest';
import { pingCommand } from '../src/commands/ping.js';
import { args, fakeClientFactory } from './helpers.js';

const NO_ARGS = args();

const fakeClient = fakeClientFactory({
  ping: vi.fn().mockResolvedValue(true),
  listNotebooks: vi.fn().mockResolvedValue([]),
});

describe('ping', () => {
  it('reports unreachable and failed auth without attempting an authenticated call', async () => {
    const listNotebooks = vi.fn();
    const client = fakeClient({ ping: vi.fn().mockResolvedValue(false), listNotebooks });

    const out = await pingCommand.run(NO_ARGS, client);

    expect(out).toContain('clipper: unreachable');
    expect(out).toContain('auth: failed');
    expect(listNotebooks).not.toHaveBeenCalled();
  });

  it('reports reachable and auth ok when an authenticated call succeeds', async () => {
    const client = fakeClient({ listNotebooks: vi.fn().mockResolvedValue([]) });

    const out = await pingCommand.run(NO_ARGS, client);

    expect(out).toContain('clipper: reachable');
    expect(out).toContain('auth: ok');
  });

  it('reports reachable but failed auth when the token is rejected', async () => {
    const client = fakeClient({ listNotebooks: vi.fn().mockRejectedValue(new Error('401')) });

    const out = await pingCommand.run(NO_ARGS, client);

    expect(out).toContain('clipper: reachable');
    expect(out).toContain('auth: failed');
  });
});
